File: src/medium/Q0005.cpp
// 5. Longest Palindromic Substring
// https://leetcode.com/problems/longest-palindromic-substring/

#include <string>

using namespace std;

// Time: O(n^2), Space: O(n), where n is the length of `s`.
auto longest_palindrome(string s) -> string {
  string longest = "";
  // Check all odd palindromes
  for (int i = 0; i < s.size(); ++i) {
    int l = i, r = i;
    while (l > -1 and r < s.size() and s[l] == s[r]) {
      const int current_len = r - l + 1;
      if (longest.size() < current_len) {
        longest = s.substr(l, current_len);
      }
      --l, ++r;
    }
  }
  // Check all even palindromes
  for (int i = 0; i < s.size() - 1; ++i) {
    int l = i, r = i + 1;
    while (l > -1 and r < s.size() and s[l] == s[r]) {
      const int current_len = r - l + 1;
      if (longest.size() < current_len) {
        longest = s.substr(l, current_len);
      }
      --l, ++r;
    }
  }
  return longest;
}

File: src/main/java/medium/Q0005.java
package medium;

/**
 * 5. Longest Palindromic Substring
 * https://leetcode.com/problems/longest-palindromic-substring/
 */
class Q0005 {

  String findLongest(int l, int r, String s, String longest) {
    while (l >= 0 && r < s.length() && s.charAt(l) == s.charAt(r)) {
      if (s.substring(l, r + 1).length() > longest.length()) {
        longest = s.substring(l, r + 1);
      }
      l--;
      r++;
    }
    return longest;
  }

  // O(n^2) time, O(n) space, where n is the length of `s`
  String longestPalindrome(String s) {
    String longest = "";
    for (int i = 0; i < s.length(); i++) {
      // Check odd length palindromes
      longest = findLongest(i, i, s, longest);
      // Check even length palindromes
      longest = findLongest(i, i + 1, s, longest);
    }
    return longest;
  }
}
